Unity-Technologies/ScriptableRenderPipeline
# Delay discard of fragments during decal projection

Metal Shading Language declares that fragment discard invalidates
derivatives for the rest of the quad, so we need to reorder when
we discard during decal projection, or we get artifacts along the
edges of the projection (any partial quads get bad partial derivatives
regardless of whether they are computed implicitly or explicitly).

**DONT FORGET TO ADD A CHANGELOG**

### Checklist for PR maker
- Have you added a Label? : HDRP, Universal, ShaderGraph etc...
- Have you added a label for backport (if needed)? : need-backport-2019.3  .  When the PR is backported the label will be change ton backported-2019.3
- Have you added a changelog? Each package have a changelog.
- Have you updated or added the documentation for you PR? When property name is changed, when a feature behavior is change, when adding a new features, think to update the documentation in the same PR.
- Have you added a graphic test for your PR (if needed)? When adding new feature or discovering a bug that isn't cover by a test, please add a graphic test

---
### Purpose of this PR
Why is this PR needed, what hard problem is it solving/fixing?

See the description.  MacOS may not encounter this problem on all platforms, but there is no guarantee as it is very driver (and/or version) dependent.  The specification of this feature is, however, unambiguous for _all_ Metal platforms.  It is also worth mentioning that without a special extension (VK_EXT_shader_demote_to_helper_invocation), Vulkan will also have this behavior by default (unclear which GPU vendors actually support it).

---
### Testing status

**Manual Tests**: What did you do?
- [x ] Opened test project + Run graphic tests locally
- [x ] Built a player
- [ x] C# and shader warnings (supress shader cache to see them)
- Other: 

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline

Any test projects to go with this to help reviewers?

The HDRP Shader Graph sample scene was used to find this problem and confirm the fix.  The decal projection (puddle) will show a "seam" artifact around the edge of the projection volume if the issue is present (so the scene should look "normal")

---
### Comments to reviewers
Notes for the reviewers you have assigned.

N.b., Much of the diffs here are based upon the whitespace config of the checked-in file.  Actual diffs are much smaller.
